test/authController.test.js

```diff
@@ -0,0 +1,108 @@
+const { app, connectDB } = require('../src/expressServer.js');
+const { MongoMemoryServer } = require('mongodb-memory-server');
+const request = require('supertest');
+const { intervalId: orderUpdateInterval } = require('../src/models/Order');
+
+// Declare variables for the server, database connection, and in-memory MongoDB instance
+let server;
+let mongooseConnection;
+let mongodb;
+
+// Credentials for test user
+const testUserCredentials = {
+  email: 'ava@ava.com',
+  password: 'secret',
+};
+
+// Test user data
+const testUser = {
+  name: 'ava',
+  username: 'ava',
+  email: 'ava@ava.com',
+  password: 'secret',
+  role: 'user',
+};
+
+// log in and return a cookie
+const loginAndReturnCookie = async (credentials) => {
+  // Send a login request and get the signed cookie from the response
+  const resp = await request(app).post('/api/v1/auth/login').send(credentials);
+
+  // Get the 'set-cookie' header from the response
+  const cookieHeader = resp.headers['set-cookie'];
+  // Find the cookie that starts with 'token' in the 'set-cookie' header
+  // Split the cookie string by ';' and take the first part (before the first ';')
+  const signedCookie = cookieHeader
+    .find((cookie) => cookie.startsWith('token'))
+    .split(';')[0];
+  return signedCookie; // Return the signed cookie, which typically contains authentication token
+};
+
+// set up the mongodb and the express server before starting the tests
+beforeAll(async () => {
+  // This will create a new instance of "MongoMemoryServer" and automatically start it
+  mongodb = await MongoMemoryServer.create();
+  const url = mongodb.getUri();
+  // set the url so that our server's mongoose connects to the in-memory mongodb and not our real one
+  process.env.MONGO_URL = url;
+  mongooseConnection = await connectDB(url);
+  server = await app.listen(8001);
+});
+
+afterAll(async () => {
+  // turn off the server and mongo connections once all the tests are done
+  await server.close();
+  await mongooseConnection.disconnect();
+  await mongodb.stop();
+  // turn off the order update interval so that jest can cleanly shutdown
+  clearInterval(orderUpdateInterval);
+});
+
+describe('Authentication API Endpoints', () => {
+  it('should register a new user and log in', async () => {
+    // Arrange: Register a new user
+    const registrationResponse = await request(app)
+      .post('/api/v1/auth/register')
+      .send(testUser);
+
+    // Assert: Check the response status and body
+    expect(registrationResponse.status).toBe(201); // Expecting a successful registration
+    expect(registrationResponse.body).toHaveProperty('user'); // Expecting a user object in the response
```

Comment:
This is great; we can also make further assertions to (1) verify that the registered user was actually stored in the db as expected, and (2) that the shape of the response from the `register` endpoint is what we expect

```suggestion
    // Assert: Check the response status and body
    expect(registrationResponse.status).toBe(201); // Expecting a successful registration
    expect(registrationResponse.body).toHaveProperty('user'); // Expecting a user object in the response
    const createdUser = await User.findOne({ email: testUser.email });
    // Expecting the user to be saved in the database (not null)
    expect(createdUser).not.toBeNull();
    // Expect the user object in the response to match the user object in the database
    expect(registrationResponse.body.user).toMatchObject({
      email: createdUser.email,
      name: createdUser.name,
      role: createdUser.role,
      userId: createdUser.id,
    });
```

Reply:
<img width="743" alt="jest" src="https://github.com/Code-the-Dream-School/dd-prac-team4-back/assets/95896159/0c1d0b26-7f1d-47e4-8a12-2a2e33b417db">